### ISSUE TITLE
[RLlib] [Bug] LSTM returns incorrect recurrent states

### ISSUE THREAD

**Issue by smorad** [ray-project/ray#21392]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### What happened + What you expected to happen

Padding affects the recurrent states that are returned from `LSTMWrapper.Forward`. LSTM should use the recurrent state produced at `seq_lens[i]`, rather than the recurrent state at `seq_lens.max()`.

E.g., we have two batches with `seq_lens = [2, 5]`. `input_dict['obs']` is zero-padded to shape `(2,5,feature)`. Then, the inputs and outputs for the LSTM are:
```
obs[b0,t0], obs[b0,t1], 0, 0, 0                            => LSTM => (b0h0, b0c0), (b0h1, b0c1), (b0h2, b0c2), (b0h3, b0c3), (b0h4, b0c4)
obs[b1,t0], obs[b1,t1], obs[b1,t2], obs[b1,t3], obs[b1,t4] => LSTM => (b1h0, b1c0), (b1h1, b1c1), (b1h2, b1c2), (b1h3, b1c3), (b1h4, b1c4)
```
At the end, we return `state = [(b0h4, b0c4), (b1h4, b1c4)]`. We should actually return `state = [(b0h1, b0c1), (b1h4, b1c4)]`. In the current implementation, the zeroth batch state results from feeding three extra zeros to the LSTM.

The error is located here: https://github.com/ray-project/ray/blob/94af7ccc9268a0ebbbfb6b79157f1f42d5925657/rllib/models/torch/recurrent_net.py#L222-L232

I think we can fix this using `torch.nn.utils.rnn.pack_padded_sequence`. I'm starting at Ray later this month, so if you can wait a few weeks I will fix this myself.

### Versions / Dependencies

Latest pip release

### Reproduction script

N/A

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!
